File: app/src/main/java/com/ramez/shopp/Adapter/SuggestedProductAdapter.java
package com.ramez.shopp.Adapter;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;

import com.ramez.shopp.Activities.ProductDetailsActivity;
import com.ramez.shopp.Activities.RegisterLoginActivity;
import com.ramez.shopp.ApiHandler.DataFeacher;
import com.ramez.shopp.Classes.AnalyticsHandler;
import com.ramez.shopp.Classes.Constants;
import com.ramez.shopp.Classes.GlobalData;
import com.ramez.shopp.Classes.MessageEvent;
import com.ramez.shopp.Classes.UtilityApp;
import com.ramez.shopp.Dialogs.CheckLoginDialog;
import com.ramez.shopp.Models.CartProcessModel;
import com.ramez.shopp.Models.LocalModel;
import com.ramez.shopp.Models.ProductBarcode;
import com.ramez.shopp.Models.ProductModel;
import com.ramez.shopp.R;
import com.ramez.shopp.Utils.NumberHandler;
import com.ramez.shopp.databinding.RowSuggestedProductItemBinding;
import com.squareup.picasso.Picasso;

import org.greenrobot.eventbus.EventBus;

import java.text.DecimalFormat;
import java.util.ArrayList;

public class SuggestedProductAdapter extends RecyclerView.Adapter<SuggestedProductAdapter.Holder> {
    //    public int count = 1;
    private Context context;
    private OnItemClick onItemClick;
    private ArrayList<ProductModel> productModels;
    private double discount = 0.0;
    private String currency = "BHD";
    private int limit = 2;
    int fraction = 2;
    LocalModel localModel;


    public SuggestedProductAdapter(Context context, ArrayList<ProductModel> productModels, OnItemClick onItemClick, int limit) {
        this.context = context;
        this.onItemClick = onItemClick;
        this.productModels = productModels;
        this.limit = limit;

    }

    @Override
    public Holder onCreateViewHolder(ViewGroup parent, int viewType) {

        RowSuggestedProductItemBinding itemView = RowSuggestedProductItemBinding.inflate(LayoutInflater.from(context), parent, false);
        return new Holder(itemView);
    }

    @SuppressLint("SetTextI18n")
    @Override
    public void onBindViewHolder(Holder holder, int position) {
        ProductModel productModel = productModels.get(position);
         localModel = UtilityApp.getLocalData() != null ? UtilityApp.getLocalData() : UtilityApp.getDefaultLocalData(context);
        currency = localModel.getCurrencyCode();
        fraction = localModel.getFractional();

        holder.binding.productNameTv.setText(productModel.getProductName().trim());

        if (productModel != null && productModel.isFavourite()) {
            holder.binding.favBut.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.favorite_icon));
        } else {
            holder.binding.favBut.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.empty_fav));

        }

        int quantity = productModel.getFirstProductBarcodes().getCartQuantity();

        if (quantity > 0) {

            holder.binding.productCartQTY.setText(String.valueOf(quantity));
            holder.binding.CartLy.setVisibility(View.VISIBLE);
            holder.binding.cartBut.setVisibility(View.GONE);

            if (quantity == 1) {
                holder.binding.deleteCartBtn.setVisibility(View.VISIBLE);
                holder.binding.minusCartBtn.setVisibility(View.GONE);
            } else {
                holder.binding.minusCartBtn.setVisibility(View.VISIBLE);
                holder.binding.deleteCartBtn.setVisibility(View.GONE);
            }

        } else {
            holder.binding.CartLy.setVisibility(View.GONE);
            holder.binding.cartBut.setVisibility(View.VISIBLE);
        }

        if (productModel.getFirstProductBarcodes().isSpecial()) {

            double originalPrice = productModel.getFirstProductBarcodes().getPrice();
            double specialPrice = productModel.getFirstProductBarcodes().getSpecialPrice();


            holder.binding.productPriceBeforeTv.setBackground(ContextCompat.getDrawable(context, R.drawable.itlatic_red_line));

            holder.binding.productPriceBeforeTv.setText(NumberHandler.formatDouble(originalPrice,
                    fraction) + " " + currency);
            holder.binding.productPriceTv.setText(NumberHandler.formatDouble(specialPrice,
                    fraction) + " " + currency);

            double discountValue = originalPrice - specialPrice;
            double discountPercent = (discountValue / originalPrice) * 100;

            if (originalPrice > 0) {
                holder.binding.discountTv.setText(NumberHandler.arabicToDecimal((int) discountPercent + " % " + "OFF"));

            } else {
                holder.binding.discountTv.setText(NumberHandler.arabicToDecimal((int) 0 + " % " + "OFF"));

            }

        } else {

            holder.binding.productPriceTv.setText(NumberHandler.formatDouble(productModel.getFirstProductBarcodes().getPrice(), fraction) + " " + currency + "");
            holder.binding.productPriceBeforeTv.setVisibility(View.GONE);
            holder.binding.discountTv.setVisibility(View.GONE);


        }

        String photoUrl = "";

        if (productModel.getImages() != null && productModel.getImages().get(0) != null && !productModel.getImages().get(0).isEmpty()) {
            photoUrl = productModel.getImages().get(0);
        } else {
            photoUrl = "http";
        }
        Picasso.get().load(photoUrl).placeholder(R.drawable.holder_image).error(R.drawable.holder_image).into(holder.binding.productImg);

//        Picasso.get().load(productModel.getImages().get(0)).placeholder(R.drawable.holder_image).error(R.drawable.holder_image).into(holder.binding.productImg);


    }

    @Override
    public int getItemCount() {
        if (limit != 0) return Math.min(productModels.size(), limit);
        else return productModels.size();


    }

    private void addToFavorite(View v, int position, int productId, int userId, int storeId) {
        AnalyticsHandler.AddToWishList(productId, currency, productId);

        new DataFeacher(false, (obj, func, IsSuccess) -> {
            if (func.equals(Constants.ERROR)) {
                GlobalData.errorDialogWithButton(context, context.getString(R.string.error),
                        context.getString(R.string.fail_to_add_favorite));

            } else if (func.equals(Constants.FAIL)) {

                GlobalData.errorDialogWithButton(context, context.getString(R.string.error),
                        context.getString(R.string.fail_to_add_favorite));
            } else {
                if (IsSuccess) {

                    Toast.makeText(context, context.getString(R.string.success_add), Toast.LENGTH_SHORT).show();

                    productModels.get(position).setFavourite(true);
                    notifyItemChanged(position);
                    notifyDataSetChanged();

                } else {

                    GlobalData.errorDialogWithButton(context, context.getString(R.string.error),
                            context.getString(R.string.fail_to_add_favorite));
                }
            }

        }).addToFavoriteHandle(userId, storeId, productId);
    }

    private void removeFromFavorite(View view, int position, int productId, int userId, int storeId) {
        new DataFeacher(false, (obj, func, IsSuccess) -> {
            if (func.equals(Constants.ERROR)) {


                GlobalData.errorDialogWithButton(context, context.getString(R.string.error),
                        context.getString(R.string.fail_to_remove_favorite));

            } else if (func.equals(Constants.FAIL)) {
                GlobalData.errorDialogWithButton(context, context.getString(R.string.error),
                        context.getString(R.string.fail_to_remove_favorite));

            } else {
                if (IsSuccess) {

                    productModels.get(position).setFavourite(false);
                    initSnackBar(context.getString(R.string.success_delete), view);
                    notifyItemChanged(position);
                    notifyDataSetChanged();


                } else {

                    GlobalData.errorDialogWithButton(context, context.getString(R.string.error),
                            context.getString(R.string.fail_to_remove_favorite));

                }
            }

        }).deleteFromFavoriteHandle(userId, storeId, productId);
    }

    private void loginFirst() {
        Toast.makeText(context, context.getString(R.string.textLoginFirst), Toast.LENGTH_SHORT).show();
        Intent intent = new Intent(context, RegisterLoginActivity.class);
        intent.putExtra(Constants.LOGIN, true);
        context.startActivity(intent);

    }

    private void initSnackBar(String message, View viewBar) {
        Toast.makeText(context, message, Toast.LENGTH_SHORT).show();

    }

    public interface OnItemClick {
        void onItemClicked(int position, ProductModel productModel);
    }

    public class Holder extends RecyclerView.ViewHolder implements View.OnClickListener {

        RowSuggestedProductItemBinding binding;

        Holder(RowSuggestedProductItemBinding view) {
            super(view.getRoot());
            binding = view;
            itemView.setOnClickListener(this);

            binding.favBut.setOnClickListener(view1 -> {

                if (!UtilityApp.isLogin()) {

                    CheckLoginDialog checkLoginDialog = new CheckLoginDialog(context, R.string.LoginFirst, R.string.to_add_favorite, R.string.ok, R.string.cancel, null, null);
                    checkLoginDialog.show();

                } else {
                    int position = getBindingAdapterPosition();
                    int userId = UtilityApp.getUserData().getId();
                    int storeId = Integer.parseInt(localModel.getCityId());
                    int productId = productModels.get(position).getId();
                    boolean isFavorite = productModels.get(position).isFavourite();
                    if (isFavorite) {
                        removeFromFavorite(view1, position, productId, userId, storeId);

                    } else {
                        addToFavorite(view1, position, productId, userId, storeId);

                    }

                }

            });

            binding.cartBut.setOnClickListener(view1 -> {

                if (!UtilityApp.isLogin()) {

                    CheckLoginDialog checkLoginDialog = new CheckLoginDialog(context, R.string.LoginFirst, R.string.to_add_cart, R.string.ok, R.string.cancel, null, null);
                    checkLoginDialog.show();

                } else {
                    String message;

                    int position = getBindingAdapterPosition();

                    ProductModel productModel = productModels.get(position);
                    ProductBarcode productBarcode = productModel.getFirstProductBarcodes();
                    int count = productBarcode.getCartQuantity();

                    int userId = UtilityApp.getUserData().getId();
                    int storeId = Integer.parseInt(localModel.getCityId());
                    int productId = productModel.getId();
                    int product_barcode_id = productBarcode.getId();
                    int limit = productBarcode.getLimitQty();
                    int stock = productBarcode.getStockQty();


                    Log.i("limit", "Log limit  " + limit);
                    Log.i("stock", "Log stock  " + stock);

                    if (limit == 0) {

                        if (count + 1 <= stock) {
                            addToCart(view1, position, productId, product_barcode_id, count + 1, userId, storeId);
                        } else {
                            message = context.getString(R.string.stock_empty);
                            GlobalData.errorDialogWithButton(context, context.getString(R.string.error),
                                    message);
                        }
                    } else {

                        if (count + 1 <= stock && (count + 1) <= limit) {
                            addToCart(view1, position, productId, product_barcode_id, count + 1, userId, storeId);
                        } else {

                            if (count + 1 > stock) {
                                message = context.getString(R.string.limit) + "" + limit;

                            } else {
                                message = context.getString(R.string.stock_empty);

                            }
                            GlobalData.errorDialogWithButton(context, context.getString(R.string.error),
                                    message);

                        }

                    }


                }

            });

            binding.plusCartBtn.setOnClickListener(v -> {
                String message = "";
                int position = getBindingAdapterPosition();

                ProductModel productModel = productModels.get(position);
                ProductBarcode productBarcode = productModel.getFirstProductBarcodes();
                int count = productBarcode.getCartQuantity();

                int userId = UtilityApp.getUserData().getId();
                int storeId = Integer.parseInt(localModel.getCityId());
                int productId = productModel.getId();
                int product_barcode_id = productBarcode.getId();
                int cart_id = productBarcode.getCartId();
                int stock = productBarcode.getStockQty();
                int limit = productBarcode.getLimitQty();

                Log.i("limit", "Log limit  " + limit);
                Log.i("stock", "Log stock  " + stock);


                if (limit == 0) {

                    if (count + 1 <= stock) {
                        updateCart(v, position, productId, product_barcode_id, count + 1, userId, storeId, cart_id, "quantity");

                    } else {
                        message = context.getString(R.string.stock_empty);
                        GlobalData.errorDialogWithButton(context, context.getString(R.string.error),
                                message);

                    }
                } else {

                    if (count + 1 <= stock && (count + 1) <= limit) {
                        updateCart(v, position, productId, product_barcode_id, count + 1, userId, storeId, cart_id, "quantity");

                    } else {
                        if (count + 1 > stock) {
                            message = context.getString(R.string.limit) + "" + limit;

                        } else {
                            message = context.getString(R.string.stock_empty);

                        }
                        GlobalData.errorDialogWithButton(context, context.getString(R.string.error),
                                message);
                    }


                }

            });

            binding.minusCartBtn.setOnClickListener(v -> {

                int position = getBindingAdapterPosition();

                ProductModel productModel = productModels.get(position);
                ProductBarcode productBarcode = productModel.getFirstProductBarcodes();
                int count = productBarcode.getCartQuantity();
                int userId = UtilityApp.getUserData().getId();
                int storeId = Integer.parseInt(localModel.getCityId());
                int productId = productModel.getId();
                int product_barcode_id = productBarcode.getId();
                int cart_id = productBarcode.getCartId();


                updateCart(v, position, productId, product_barcode_id, count - 1, userId, storeId, cart_id, "quantity");


            });

            binding.deleteCartBtn.setOnClickListener(v -> {
                int position = getBindingAdapterPosition();
                ProductModel productModel = productModels.get(position);
                ProductBarcode productBarcode = productModel.getFirstProductBarcodes();
                onItemClick.onItemClicked(position, productModels.get(position));

                int userId = UtilityApp.getUserData().getId();
                int storeId = Integer.parseInt(localModel.getCityId());
                int productId = productModel.getId();
                int product_barcode_id = productBarcode.getId();
                int cart_id = productBarcode.getCartId();

                deleteCart(v, position, productId, product_barcode_id, cart_id, userId, storeId);

            });


        }


        @Override
        public void onClick(View v) {

            if (onItemClick != null) {
                if (productModels.size() > 0) {
                    int position = getBindingAdapterPosition();
                    onItemClick.onItemClicked(position, productModels.get(position));
                    ProductModel productModel = productModels.get(position);
                    Intent intent = new Intent(context, ProductDetailsActivity.class);
                    intent.putExtra(Constants.DB_productModel, productModel);
                    context.startActivity(intent);
                }


            }
        }


        private void addToCart(View v, int position, int productId, int product_barcode_id, int quantity, int userId, int storeId) {


            if (quantity > 0) {
                new DataFeacher(false, (obj, func, IsSuccess) -> {
                    CartProcessModel result = (CartProcessModel) obj;

                    if (IsSuccess) {
                        int cartId = result.getId();

                        productModels.get(position).getFirstProductBarcodes().setCartQuantity(quantity);
                        productModels.get(position).getFirstProductBarcodes().setCartId(cartId);
                        notifyItemChanged(position);
                        System.out.println("Log suggest addToCart" + result.getTotal());
                        UtilityApp.updateCart(1, productModels.size());

                        AnalyticsHandler.AddToCart(result.getId(), currency, quantity);


                        EventBus.getDefault().post(new MessageEvent(MessageEvent.TYPE_READ_CART));


                    } else {
                        GlobalData.errorDialogWithButton(context, context.getString(R.string.error),
                                context.getString(R.string.fail_to_add_cart));


                    }

                }).addCartHandle(productId, product_barcode_id, quantity, userId, storeId);


            } else {
                Toast.makeText(context, context.getString(R.string.quanity_wrong), Toast.LENGTH_SHORT).show();
            }

        }

        private void updateCart(View v, int position, int productId, int product_barcode_id, int quantity, int userId, int storeId, int cart_id, String update_quantity) {

            if (quantity > 0) {
                new DataFeacher(false, (obj, func, IsSuccess) -> {
                    if (IsSuccess) {
//                    initSnackBar(context.getString(R.string.success_to_update_cart), v);
                        productModels.get(position).getFirstProductBarcodes().setCartQuantity(quantity);
                        notifyItemChanged(position);

                    } else {
                        GlobalData.errorDialogWithButton(context, context.getString(R.string.error),
                                context.getString(R.string.fail_to_update_cart));


                    }

                }).updateCartHandle(productId, product_barcode_id, quantity, userId, storeId, cart_id, update_quantity);


            } else {
                Toast.makeText(context, context.getString(R.string.quanity_wrong), Toast.LENGTH_SHORT).show();
            }

        }

        private void deleteCart(View v, int position, int productId, int product_barcode_id, int cart_id, int userId, int storeId) {
            new DataFeacher(false, (obj, func, IsSuccess) -> {

                if (IsSuccess) {
                    productModels.get(position).getFirstProductBarcodes().setCartQuantity(0);
                    notifyItemChanged(position);
                    initSnackBar(context.getString(R.string.success_delete_from_cart), v);
                    UtilityApp.updateCart(2, productModels.size());
                    EventBus.getDefault().post(new MessageEvent(MessageEvent.TYPE_READ_CART));
                    AnalyticsHandler.RemoveFromCart(cart_id, currency, 0);


                } else {

                    GlobalData.errorDialogWithButton(context, context.getString(R.string.delete_product),
                            context.getString(R.string.fail_to_delete_cart));
                }


            }).deleteCartHandle(productId, product_barcode_id, cart_id, userId, storeId);
        }

    }

}

File: app/src/main/java/com/ramez/shopp/adapter/SuggestedProductAdapter.kt
package com.ramez.shopp.adapter


import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.ramez.shopp.ApiHandler.DataFeacher
import com.ramez.shopp.ApiHandler.DataFetcherCallBack
import com.ramez.shopp.Classes.*
import com.ramez.shopp.Dialogs.CheckLoginDialog
import com.ramez.shopp.Models.CartProcessModel
import com.ramez.shopp.Models.LocalModel
import com.ramez.shopp.Models.ProductModel
import com.ramez.shopp.R
import com.ramez.shopp.Utils.NumberHandler
import com.ramez.shopp.activities.ProductDetailsActivity
import com.ramez.shopp.databinding.RowSuggestedProductItemBinding
import org.greenrobot.eventbus.EventBus
import java.lang.Exception
import java.util.ArrayList


class SuggestedProductAdapter(
    private val context: Context,
    private val productModels: ArrayList<ProductModel>?,
    private val onItemClick: OnItemClick?,
    limit: Int
) :
        RecyclerView.Adapter<SuggestedProductAdapter.Holder>() {
    private var currency = "BHD"
    private var limit = 2
    var fraction = 2
    var localModel: LocalModel? = null
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): Holder {
        val itemView = RowSuggestedProductItemBinding.inflate(
            LayoutInflater.from(
                context
            ), parent, false
        )
        return Holder(itemView)
    }

    @SuppressLint("SetTextI18n")
    override fun onBindViewHolder(holder: Holder, position: Int) {
        val productModel = productModels!![position]
        localModel =
            if (UtilityApp.getLocalData() != null) UtilityApp.getLocalData() else UtilityApp.getDefaultLocalData(
                context
            )
        currency = localModel?.currencyCode ?:Constants.BHD
        fraction = localModel?.fractional ?:Constants.three
        holder.binding.productNameTv.text = productModel.productName.trim { it <= ' ' }
        if (productModels.size > 0 && productModel.isFavourite) {
            holder.binding.favBut.setImageDrawable(
                ContextCompat.getDrawable(
                    context,
                    R.drawable.favorite_icon
                )
            )
        } else {
            holder.binding.favBut.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.empty_fav))
        }
        val quantity = productModel.firstProductBarcodes.cartQuantity
        if (quantity > 0) {
            holder.binding.productCartQTY.text = quantity.toString()
            holder.binding.CartLy.visibility = View.VISIBLE
            holder.binding.cartBut.visibility = View.GONE
            if (quantity == 1) {
                holder.binding.deleteCartBtn.visibility = View.VISIBLE
                holder.binding.minusCartBtn.visibility = View.GONE
            } else {
                holder.binding.minusCartBtn.visibility = View.VISIBLE
                holder.binding.deleteCartBtn.visibility = View.GONE
            }
        } else {
            holder.binding.CartLy.visibility = View.GONE
            holder.binding.cartBut.visibility = View.VISIBLE
        }
        if (productModel.firstProductBarcodes.isSpecial) {
            val originalPrice = productModel.firstProductBarcodes.price
            val specialPrice = productModel.firstProductBarcodes.specialPrice
            holder.binding.productPriceBeforeTv.background = ContextCompat.getDrawable(
                context, R.drawable.itlatic_red_line
            )
            holder.binding.productPriceBeforeTv.text = NumberHandler.formatDouble(
                originalPrice,
                fraction
            ) + " " + currency
            holder.binding.productPriceTv.text = NumberHandler.formatDouble(
                specialPrice,
                fraction
            ) + " " + currency
            val discountValue = originalPrice - specialPrice
            val discountPercent = (discountValue.div(originalPrice)) .times(100)


            if (originalPrice > 0) {
                holder.binding.discountTv.text =
                    NumberHandler.arabicToDecimal(discountPercent.toInt().toString()).plus( " % " + "OFF")
            } else {
                holder.binding.discountTv.text =
                    NumberHandler.arabicToDecimal("0").plus( " % " + "OFF")
            }
        } else {
            holder.binding.productPriceTv.text =
                NumberHandler.formatDouble(
                    productModel.firstProductBarcodes.price,
                    fraction
                ) + " " + currency + ""
            holder.binding.productPriceBeforeTv.visibility = View.GONE
            holder.binding.discountTv.visibility = View.GONE
        }
        var photoUrl: String? = ""
        photoUrl =
            if (productModel.images != null && productModel.images[0] != null && !productModel.images[0].isEmpty()) {
                productModel.images[0]
            } else {
                "http"
            }
        try {
            GlobalData.GlideImg(
                context, photoUrl, R.drawable.holder_image, holder.binding.productImg
            )
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    override fun getItemCount(): Int {
        return if (limit != 0) Math.min(productModels!!.size, limit) else productModels!!.size
    }

    private fun addToFavorite(v: View, position: Int, productId: Int, userId: Int, storeId: Int) {
        AnalyticsHandler.AddToWishList(productId, currency, productId.toDouble())
        DataFeacher(false,
            object :DataFetcherCallBack {
                override fun Result(obj: Any?, func: String?, IsSuccess: Boolean) {
                    if (func == Constants.ERROR) {
                        GlobalData.errorDialogWithButton(
                            context, context.getString(R.string.error),
                            context.getString(R.string.fail_to_add_favorite)
                        )
                    } else if (func == Constants.FAIL) {
                        GlobalData.errorDialogWithButton(
                            context, context.getString(R.string.error),
                            context.getString(R.string.fail_to_add_favorite)
                        )
                    } else {
                        if (IsSuccess) {
                            Toast.makeText(context, context.getString(R.string.success_add), Toast.LENGTH_SHORT)
                                .show()
                            productModels!![position].isFavourite = true
                            notifyItemChanged(position)
                            notifyDataSetChanged()
                        } else {
                            GlobalData.errorDialogWithButton(
                                context, context.getString(R.string.error),
                                context.getString(R.string.fail_to_add_favorite)
                            )
                        }
                    }                }

            }).addToFavoriteHandle(userId, storeId, productId)
    }

    private fun removeFromFavorite(view: View, position: Int, productId: Int, userId: Int, storeId: Int) {
        DataFeacher(false,object :
            DataFetcherCallBack {
            override fun Result(obj: Any?, func: String?, IsSuccess: Boolean) {
                if (func == Constants.ERROR) {
                    GlobalData.errorDialogWithButton(
                        context, context.getString(R.string.error),
                        context.getString(R.string.fail_to_remove_favorite)
                    )
                } else if (func == Constants.FAIL) {
                    GlobalData.errorDialogWithButton(
                        context, context.getString(R.string.error),
                        context.getString(R.string.fail_to_remove_favorite)
                    )
                } else {
                    if (IsSuccess) {
                        productModels!![position].isFavourite = false
                        initSnackBar(context.getString(R.string.success_delete), view)
                        notifyItemChanged(position)
                        notifyDataSetChanged()
                    } else {
                        GlobalData.errorDialogWithButton(
                            context, context.getString(R.string.error),
                            context.getString(R.string.fail_to_remove_favorite)
                        )
                    }
                }            }

        }).deleteFromFavoriteHandle(userId, storeId, productId)
    }

    private fun initSnackBar(message: String, viewBar: View) {
        Toast.makeText(context, message, Toast.LENGTH_SHORT).show()
    }

    interface OnItemClick {
        fun onItemClicked(position: Int, productModel: ProductModel?)
    }

    inner class Holder internal constructor(var binding: RowSuggestedProductItemBinding) :
            RecyclerView.ViewHolder(
                binding.root
            ),
            View.OnClickListener {
        override fun onClick(v: View) {
            if (onItemClick != null) {
                if (productModels != null && productModels.size > 0) {
                    val position = bindingAdapterPosition
                    onItemClick.onItemClicked(position, productModels[position])
                    val productModel = productModels[position]
                    val intent = Intent(context, ProductDetailsActivity::class.java)
                    intent.putExtra(Constants.DB_productModel, productModel)
                    context.startActivity(intent)
                }
            }
        }

        private fun addToCart(
            v: View,
            position: Int,
            productId: Int,
            product_barcode_id: Int,
            quantity: Int,
            userId: Int,
            storeId: Int
        ) {
            if (quantity > 0) {
                DataFeacher(false, object :DataFetcherCallBack {
                    override fun Result(obj: Any?, func: String?, IsSuccess: Boolean) {
                        val result = obj as CartProcessModel
                        if (IsSuccess) {
                            val cartId = result.id
                            productModels?.get(position)?.firstProductBarcodes?.cartQuantity  = quantity
                            productModels?.get(position)?.firstProductBarcodes?.cartId = cartId
                            notifyItemChanged(position)
                            println("Log suggest addToCart" + result.total)
                            if (productModels?.size ?: 0 > 0) {
                                UtilityApp.updateCart(1, productModels?.size?:0)
                            }
                            AnalyticsHandler.AddToCart(result.id, currency, quantity.toDouble())
                            EventBus.getDefault()
                                .post(MessageEvent(MessageEvent.TYPE_READ_CART))
                        } else {
                            GlobalData.errorDialogWithButton(
                                context, context.getString(R.string.error),
                                context.getString(R.string.fail_to_add_cart)
                            )
                        }                    }

                }).addCartHandle(productId, product_barcode_id, quantity, userId, storeId)
            } else {
                Toast.makeText(context, context.getString(R.string.quanity_wrong), Toast.LENGTH_SHORT).show()
            }
        }

        private fun updateCart(
            v: View,
            position: Int,
            productId: Int,
            product_barcode_id: Int,
            quantity: Int,
            userId: Int,
            storeId: Int,
            cart_id: Int,
            update_quantity: String
        ) {
            if (quantity > 0) {
                DataFeacher(false,
                    object :DataFetcherCallBack {
                        override fun Result(obj: Any?, func: String?, IsSuccess: Boolean) {
                            if (IsSuccess) {

                                productModels?.get(position)?.firstProductBarcodes?.cartQuantity  = quantity
                                notifyItemChanged(position)
                            } else {
                                GlobalData.errorDialogWithButton(
                                    context, context.getString(R.string.error),
                                    context.getString(R.string.fail_to_update_cart)
                                )
                            }                        }

                    }).updateCartHandle(
                    productId,
                    product_barcode_id,
                    quantity,
                    userId,
                    storeId,
                    cart_id,
                    update_quantity
                )
            } else {
                Toast.makeText(context, context.getString(R.string.quanity_wrong), Toast.LENGTH_SHORT).show()
            }
        }

        private fun deleteCart(
            v: View,
            position: Int,
            productId: Int,
            product_barcode_id: Int,
            cart_id: Int,
            userId: Int,
            storeId: Int
        ) {
            DataFeacher(false,
                object :DataFetcherCallBack {
                    override fun Result(obj: Any?, func: String?, IsSuccess: Boolean) {
                        if (IsSuccess) {
                            productModels!![position].firstProductBarcodes.cartQuantity = 0
                            notifyItemChanged(position)
                            initSnackBar(context.getString(R.string.success_delete_from_cart), v)
                            UtilityApp.updateCart(2, productModels.size)
                            EventBus.getDefault()
                                .post(MessageEvent(MessageEvent.TYPE_READ_CART))
                            AnalyticsHandler.RemoveFromCart(cart_id, currency, 0.0)
                        } else {
                            GlobalData.errorDialogWithButton(
                                context, context.getString(R.string.delete_product),
                                context.getString(R.string.fail_to_delete_cart)
                            )
                        }                    }

                }).deleteCartHandle(productId, product_barcode_id, cart_id, userId, storeId)
        }

        init {
            itemView.setOnClickListener(this)
            binding.favBut.setOnClickListener { view1 ->
                if (!UtilityApp.isLogin()) {
                    val checkLoginDialog = CheckLoginDialog(
                        context,
                        R.string.LoginFirst,
                        R.string.to_add_favorite,
                        R.string.ok,
                        R.string.cancel,
                        null,
                        null
                    )
                    checkLoginDialog.show()
                } else {
                    val position = bindingAdapterPosition
                    val userId = UtilityApp.getUserData().id
                    val storeId = localModel!!.cityId.toInt()
                    val productId = productModels!![position].id
                    val isFavorite = productModels[position].isFavourite
                    if (isFavorite) {
                        removeFromFavorite(view1, position, productId, userId, storeId)
                    } else {
                        addToFavorite(view1, position, productId, userId, storeId)
                    }
                }
            }
            binding.cartBut.setOnClickListener { view1 ->
                if (!UtilityApp.isLogin()) {
                    val checkLoginDialog = CheckLoginDialog(
                        context,
                        R.string.LoginFirst,
                        R.string.to_add_cart,
                        R.string.ok,
                        R.string.cancel,
                        null,
                        null
                    )
                    checkLoginDialog.show()
                } else {
                    val message: String
                    val position = bindingAdapterPosition
                    val productModel = productModels!![position]
                    val productBarcode = productModel.firstProductBarcodes
                    val count = productBarcode.cartQuantity
                    val userId = UtilityApp.getUserData().id
                    val storeId = localModel!!.cityId.toInt()
                    val productId = productModel.id
                    val productBarcodeId = productBarcode.id
                    val limit = productBarcode.limitQty
                    val stock = productBarcode.stockQty
                    Log.i("limit", "Log limit  $limit")
                    Log.i("stock", "Log stock  $stock")
                    if (limit == 0) {
                        if (count + 1 <= stock) {
                            addToCart(
                                view1,
                                position,
                                productId,
                                productBarcodeId,
                                count + 1,
                                userId,
                                storeId
                            )
                        } else {
                            message = context.getString(R.string.stock_empty)
                            GlobalData.errorDialogWithButton(
                                context, context.getString(R.string.error),
                                message
                            )
                        }
                    } else {
                        if (count + 1 <= stock && count + 1 <= limit) {
                            addToCart(
                                view1,
                                position,
                                productId,
                                productBarcodeId,
                                count + 1,
                                userId,
                                storeId
                            )
                        } else {
                            message = if (count + 1 > stock) {
                                context.getString(R.string.limit) + "" + limit
                            } else {
                                context.getString(R.string.stock_empty)
                            }
                            GlobalData.errorDialogWithButton(
                                context, context.getString(R.string.error),
                                message
                            )
                        }
                    }
                }
            }
            binding.plusCartBtn.setOnClickListener { v ->
                var message = ""
                val position = bindingAdapterPosition
                val productModel = productModels!![position]
                val productBarcode = productModel.firstProductBarcodes
                val count = productBarcode.cartQuantity
                val userId = UtilityApp.getUserData().id
                val storeId = localModel!!.cityId.toInt()
                val productId = productModel.id
                val productBarcodeId = productBarcode.id
                val cartId = productBarcode.cartId
                val stock = productBarcode.stockQty
                val limit = productBarcode.limitQty
                Log.i("limit", "Log limit  $limit")
                Log.i("stock", "Log stock  $stock")
                if (limit == 0) {
                    if (count + 1 <= stock) {
                        updateCart(
                            v,
                            position,
                            productId,
                            productBarcodeId,
                            count + 1,
                            userId,
                            storeId,
                            cartId,
                            "quantity"
                        )
                    } else {
                        message = context.getString(R.string.stock_empty)
                        GlobalData.errorDialogWithButton(
                            context, context.getString(R.string.error),
                            message
                        )
                    }
                } else {
                    if (count + 1 <= stock && count + 1 <= limit) {
                        updateCart(
                            v,
                            position,
                            productId,
                            productBarcodeId,
                            count + 1,
                            userId,
                            storeId,
                            cartId,
                            "quantity"
                        )
                    } else {
                        message = if (count + 1 > stock) {
                            context.getString(R.string.limit) + "" + limit
                        } else {
                            context.getString(R.string.stock_empty)
                        }
                        GlobalData.errorDialogWithButton(
                            context, context.getString(R.string.error),
                            message
                        )
                    }
                }
            }
            binding.minusCartBtn.setOnClickListener { v ->
                val position = bindingAdapterPosition
                val productModel = productModels!![position]
                val productBarcode = productModel.firstProductBarcodes
                val count = productBarcode.cartQuantity
                val userId = UtilityApp.getUserData().id
                val storeId = localModel!!.cityId.toInt()
                val productId = productModel.id
                val productBarcodeId = productBarcode.id
                val cartId = productBarcode.cartId
                updateCart(
                    v,
                    position,
                    productId,
                    productBarcodeId,
                    count - 1,
                    userId,
                    storeId,
                    cartId,
                    "quantity"
                )
            }
            binding.deleteCartBtn.setOnClickListener { v ->
                val position = bindingAdapterPosition
                val productModel = productModels!![position]
                val productBarcode = productModel.firstProductBarcodes
                onItemClick!!.onItemClicked(position, productModels[position])
                val userId = UtilityApp.getUserData().id
                val storeId = localModel!!.cityId.toInt()
                val productId = productModel.id
                val productBarcodeId = productBarcode.id
                val cartId = productBarcode.cartId
                deleteCart(v, position, productId, productBarcodeId, cartId, userId, storeId)
            }
        }
    }

    init {
        this.limit = limit
    }
}
